### ISSUE TITLE
Major: Cpu-benchmark completely misleading (single-core benchmark)

### ISSUE THREAD

**Issue by germanicianus** [FAForever/fa#2434]
The cpu-benchmark currently uses just one core. The game is multi-threaded and therefore the result is completely misleading. One of my cpus is an old one, but has 4 cores. It gets a "CPU Rating" of 373 which results in a violet bar almost fills up the benchmark bar. Nonetheless the cpu is capable of handling a 12 player online game without any problems and can even speed up a replay of a 12 player game to +5.

Players use that benchmark result to decide which players to kick from a game. If the benchmark is so unreliable like it is now, this is a big problem.

Benchmark using just a single core:
![image](https://user-images.githubusercontent.com/37223176/38107353-f81c07ee-3391-11e8-9741-9d2ceb746130.png)

Speeding 12 player replay up +7:
![image](https://user-images.githubusercontent.com/37223176/38107656-e2049e2a-3392-11e8-8da1-f6e672ae3e8b.png)

**Comment by speed2CZ** [comment]
10 minutes into the game is nothing, just from the corner screenshot I can see that even my granpa CPU could handle that number of units

There are two major threads in FA:
sim - all simulation and AI
ren - all rendering and UI
those two eat the most performance, each on different core. The rest of the threads for sounds and what not wont take more than 5%
You can try to run some pro setons replay to the point when the sim speed won't get faster than 0 or goes negative and you will see that the FA will be still around 50% of CPU usage. (as I said earlier 2 major threads, each on different core.)

**Comment by germanicianus** [comment]
Many thanks for your great explanation. I allready guessed that there are 2 major threads and the rest are minor ones. For the last screenshot I set their core affinity manually.

With your explanation you are reinforcing my point that the benchmark is currently very misleading because it should consider at least 2 cores for the 2 major threads.

On all games I had until know, my cpu capable of keeping the pace. The games also had some very good players and lasted long. I will keep an eye on that :) .

**Comment by speed2CZ** [comment]
No, the deciding factor is the sim speed. That runs on a single core. So testing a single core performance is valid.

**Comment by micheljung** [comment]
Might I add here: Single-Core performance is valid as long as the player has at least 2 cores, but I hope everyone does today :D

And if someone is saying "but why is there no core at 100%": because Windows runs threads not exclusively on one core but it rotates. If you have 4 cores, a thread will run 1/4s on core 1, 1/4s on core 2, and so on so you'll see all cores at 25%.

**Comment by ChrisKitching** [comment]
It's worth noting that the benchmark result is cached in the local profile store. So if you're someone for whom the benchmark doesn't give a realistic measure of your actual game performance, you can just edit the appropriate text file and set the stored score to something that gets you kicked less often. Not abusing this power is left as an exercise for the reader.

The benchmark is far from perfect, as its workload doesn't really match up with what the game does during play. I've definitely seen cases (including on my own machines) where it gives an extremely poor score for a machine that runs the game just fine.

The benchmark will also become completely useless if attempts to integrate LuaJIT are ever successful, since the JIT will detect that the benchmark loop is redundant and optimise it out...

Another problem with the benchmark is that executing it causes the lobby's UI to become unresponsive for several seconds. Not exactly the best experience for new players joining their first game. This UX issue could be mitigated by running the benchmark on the loading screen on first launch, before rendering the lobby UI.

A more future-proof (and accurate) solution might be to benchmark what you actually care about: the game. Clients can measure how long Lua execution took during a particular simulation timestep, and realise that they're not fast enough when this exceeds some threshold governed by the sim frequency. This would give you a very good measure of how well your computer copes with running the game, independent of network latency, and this information can be used to create a very accurate rating for people's computers. You can start making really high quality recommendations like "your computer seems to cope with games involving up to 4 players with 500 units each", and can warn users attempting to join games that their performance history suggests they can't handle. The FAF client would probably want to track hardware information and invalidate the performance history information when the hardware changes.

**Comment by FAETHER** [comment]
I used to modify my cpu benchmark to 25 or lower and people in lobby were very confused on what a beast my machine is ;). I would usually say I run 2 itel Xeon processors with 20 cores !

**Comment by micheljung** [comment]
I still suggest TrueSkill for CPU rating at the end of each game ;-)

**Comment by Uveso** [comment]
Well, i am saying this in all benchmark discussions.
CPU power is no longer our main problem. (Since we have fast CPU's like Intel I7 etc)
As already stated; If the game goes under +0 simscore then all CPU's are still under 25% load. (Intel I7 at least)
This means, we don't have a problem with calculation speed.

I did some Hardwaretest on my PCS's that convinced me, that memory (RAM) speed is way more important in these days.

So i suggest to add memory test's like tabledeepcopy etc. to get a propper PC rating.

**Comment by germanicianus** [comment]
CPU power is of course still a problem, because people expect SC to run on older computers and on mainstream laptops which it does, but not fluid with many units. [One example on Steam SC forums.](https://steamcommunity.com/app/9420/discussions/0/3288067088123968521/#c1694919808754991741)

**Comment by Uveso** [comment]
germanicianus,
i did not wrote that cpu is not a problem anymore.
I wrote it's no longer our main problem if you have a fast CPU like an Intel I7.

And i was suggesting to "ADD" a memory test for newer CPU's,
and not to replace/remove CPU test for older CPU's.

Yes you are right, people are expecting that SupCom runns on a slow PC.
But i am talking about FAST PC's :)
And that we need additional benchmark test's for FAST cpu's like memory test.

**Comment by germanicianus** [comment]
Imho a ["time demo"](http://wiki.splashdamage.com/index.php/Benchmarking#2._Multiplayer_Performance_using_.27timeNetDemo.27) of a recorded 12 player game with an extremely high number of units with turned of/down graphics and unlimited simulation speed would be the ideal solution.

**Comment by MifLegend** [comment]
> I used to modify my cpu benchmark to 25 or lower and people in lobby were very confused on what a beast my machine is ;). I would usually say I run 2 itel Xeon processors with 20 cores !

можешь написать как ты это сделал?)
давно ищу способ вручную изменить показатель проца.
